### ISSUE TITLE
the inference logic may be different from the paper?

### ISSUE THREAD

**Issue by Leiyi-HU** [GuoleiSun/VSS-CFFM#9]
`        if not self.training and num_clips!=self.num_clips: # for single frame inference.
        # if not self.training:
            return x[:,-1] # return the last clip.`
use this code, if only 1 image is put into the decoder_head, the code following will not be used, and thus, the cffm head will be same to the segformer head.
`h2=int(h/2)
        w2=int(w/2)
        _c = resize(_c, size=(h2,w2),mode='bilinear',align_corners=False)

        _c_further=_c.reshape(batch_size, num_clips, -1, h2, w2) # bdchw
        # print(_c_further.shape)
        # exit()
        _c2=self.decoder_focal(_c_further) # do focal on h2,w2 size to save computation.
        # _c_further=_c_further.permute(0,2,1,3,4)

        # _c2=_c.reshape(batch_size, num_clips, -1, h, w)

        assert _c_further.shape==_c2.shape # comments: not change the size of the features.

        _c_further2=torch.cat([_c_further[:,-1], _c2[:,-1]],1)

        x2 = self.dropout(_c_further2)
        x2 = self.linear_pred2(x2)
        x2=resize(x2, size=(h,w),mode='bilinear',align_corners=False)
        x2=x2.unsqueeze(1)

        x=torch.cat([x,x2],1)   ## b*(k+1)*124*h*w`

**Comment by Leiyi-HU** [comment]
And, I found that the code only put one image into the model when inference, I couldn't find the code which caches the previous feature maps as the paper reported. could you please tell me where is it? Thanks a lot!

**Comment by GuoleiSun** [comment]
Hi, thanks for your interest. 
The inference is like this:
1. For the first (early) frame(s) of a video, the previous frames are less than 4. So one image is forwarded to the model.
2. For other frames of a video, we input 4 frames to the model and CFFM head is used. You could run the code and check the details. In our released code, we don't provide the function of caching previous feature maps for simplicity. So 4 frames are forward to the model and the result of the target frame is saved.

**Comment by Leiyi-HU** [comment]
> 

Thanks! I have found that you input 4 frames to the inference model, so sorry to bother you. and, the fps reported in the paper was computed according to 4 frames input or 1?

**Comment by GuoleiSun** [comment]
For fps, the features for 3 previous frames are in the memory and the encoder only needs to process the target frame. Then, our CFFM head would take all features (for 4 frames) and compute the mask for the target.
